File: part-forge-i/scripts/python/pcpp_worker.py
#!/usr/bin/env python3
"""
Long-running worker that ingests PCPartPicker data on an interval.
- Uses the existing ingest_pcpartpicker.py main() each cycle
- Interval configurable via PCPP_INTERVAL_SEC (default: 900 seconds)
- Region/categories controlled by ingest script; extend as needed
"""
import os
import time
import signal
import threading

# Reuse the existing ingestion logic
import ingest_pcpartpicker as ingest

shutdown = threading.Event()


def handle_signal(signum, frame):
    print(f"[pcpp-worker] Received signal {signum}, shutting down...")
    shutdown.set()


def run_loop():
    interval = int(os.getenv("PCPP_INTERVAL_SEC", "900"))
    print(f"[pcpp-worker] Starting loop with interval={interval}s")

    while not shutdown.is_set():
        start = time.time()
        try:
            ingest.main()
            elapsed = time.time() - start
            print(f"[pcpp-worker] Ingest cycle completed in {elapsed:.1f}s")
        except Exception as e:
            print(f"[pcpp-worker] ERROR during ingest: {e}")
            # small backoff on error
            time.sleep(min(60, interval))
        # sleep remaining interval (if any)
        remaining = interval - (time.time() - start)
        if remaining > 0 and not shutdown.is_set():
            shutdown.wait(timeout=remaining)


def main():
    # Graceful shutdown hooks
    signal.signal(signal.SIGINT, handle_signal)
    signal.signal(signal.SIGTERM, handle_signal)
    run_loop()


if __name__ == "__main__":
    main()

File: part-forge-i/src/scripts/pcpp_worker.js
// Node worker that periodically runs the Python scraper
// Requires Python and pip dependencies installed at build
const { spawn } = require('child_process');
const path = require('path');

const INTERVAL_SEC = parseInt(process.env.PCPP_INTERVAL_SEC || '900', 10);
let shuttingDown = false;

process.on('SIGINT', onSignal);
process.on('SIGTERM', onSignal);

function onSignal(sig) {
  console.log(`[pcpp-node-worker] Received ${sig}, shutting down...`);
  shuttingDown = true;
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

async function runOnce() {
  return new Promise((resolve) => {
    const cwd = path.resolve(__dirname, '../../');
    console.log(`[pcpp-node-worker] Starting ingest (cwd=${cwd})`);
    const proc = spawn('python', ['scripts/python/ingest_pcpartpicker.py'], {
      cwd,
      stdio: 'inherit',
      env: process.env,
    });
    proc.on('exit', (code, signal) => {
      console.log(`[pcpp-node-worker] Ingest exited code=${code} signal=${signal}`);
      resolve({ code, signal });
    });
  });
}

async function main() {
  console.log(`[pcpp-node-worker] Launching with interval=${INTERVAL_SEC}s`);
  while (!shuttingDown) {
    const start = Date.now();
    try {
      const { code } = await runOnce();
      if (code !== 0) {
        console.error('[pcpp-node-worker] Non-zero exit from ingest, will retry after short backoff');
        await sleep(Math.min(60000, INTERVAL_SEC * 1000));
      }
    } catch (err) {
      console.error('[pcpp-node-worker] ERROR running ingest:', err);
      await sleep(Math.min(60000, INTERVAL_SEC * 1000));
    }

    const elapsed = (Date.now() - start) / 1000;
    const remaining = INTERVAL_SEC - elapsed;
    if (remaining > 0 && !shuttingDown) {
      await sleep(remaining * 1000);
    }
  }
  console.log('[pcpp-node-worker] Stopped');
}

main().catch((e) => {
  console.error('[pcpp-node-worker] Fatal error:', e);
  process.exit(1);
});